Let's Write a Qi Compiler!

countvajhula opened the pull request:
### Summary of Changes

Integration branch for the compiler work. This will eventually be merged into main when the compiler is ready.

See [Qi Compiler Meeting Notes](https://github.com/drym-org/qi/wiki/Meeting-Notes#qi-compiler).

### Work Remaining

- [ ] Investigate Qi forms still being unlinked in the docs, though they are now exported bindings (but in the Qi binding space) and required `for-label`
- [ ] Validate that `kw-helper` is being correctly used in left- and right-currying cases (see `#%blanket-template` code generation in `compiler.rkt`) [Ben/Noah?]

### WIP

- [ ] Update the Qi docs [Sid] (#146 )
- [ ] Generate final performance reports [Dominik]
- [ ] Code review

### Already Done

(reverse-chronological, most recently completed at the top)

- [x] Update the [source code layout and overview](https://github.com/drym-org/qi/wiki/Source-Code-Overview) in the wiki
- [x] Fix the marshalling issue on Racket BC 8.5 (#147 )
- [ ] Incorporate competitive benchmarks into the [live dashboard](https://drym-org.github.io/qi/benchmarks/) [Dominik/Sid?] [We're going to just have standalone benchmarks hosted somewhere for now, and will incorporate them into the SDK once it's released as a library and we've worked out the appropriate triggers for running them]
- [x] Fix the "weird bug" re: improper attempt to optimize a subexpression (#143 )
- [x] Update developer docs [on the compiler](https://github.com/drym-org/qi/wiki/The-Compiler) to describe the implementation (at a high level) and any best practices for working on it / debugging etc. [Dominik/Sid?]
- [x] Migrate the `relation` library [Sid]
- [x] Review code coverage and increase it (ideally to 100%) (#86)
- [x] Start a ["Release Notes"](https://github.com/drym-org/qi/wiki/Release-Notes) wiki including bw incompatibilities
- [x] Review use of side effects, specifically `esc` (and implicitly escaped like just using `displayln`), `effect` and `gen` should be honored (do these interact with deforestation or normalization?) [Ben/Dominik?]
- [x] Fix the "bind and switch" issue
- [x] Beat Racket on one benchmark!
- [x] Publish syntax-spec to the package index
~- [ ] Add docs for syntax-spec~
- [x] Write a compiler optimization together with a validating nonlocal benchmark (#76 )
- [x] Organize benchmarks to prepare to add nonlocal benchmarks (#95 )
- [x] Restore performance with the new architecture at least to parity with the main branch (#89 )
- [x] Do benchmarking to quantify performance regressions from the main branch (#88 )
- [x] Introduce bindings into Qi (#75 )
- [x] Incorporate [bindingspec](https://github.com/michaelballantyne/bindingspec) to abstract the expander (#68 )
- [x] Distill a small core language (#63 )
- [x] Separate expansion and compilation phases (#49 -- merged into main)

### Pre-Merge Checklist

- [ ] Create [version exceptions](https://docs.racket-lang.org/pkg/getting-started.html#%28part._.Version_.Exceptions%29) on the package index for Racket 8.3 and 8.4 (or avoid using `all-defined-out` -- see racket/racket#4154)
- [ ] Notify known dependent projects re: backwards incompatibility, and define a migration plan (e.g. for the [`relation`](https://docs.racket-lang.org/relation/index.html) library)
- [ ] Test Frosthaven Manager and other codebases using Qi with the compiler branch

### Public Domain Dedication

- [x] In contributing, we relinquish any copyright claims on our contributions and freely release them into the public domain in the simple hope that they will provide value. ([Why](https://github.com/drym-org/foundation/blob/main/Declaration_of_Non_Ownership.md))

countvajhula commented:
FYI @michaelballantyne @benknoble

countvajhula commented:
Current performance comparison of individual forms with a priori (main branch) performance (repeated the same analysis that was [done a little while back](https://github.com/drym-org/qi/pull/63#issuecomment-1220082611)):

```
'(("live?" "55.71")
  ("count" "33.4")
  ("rectify" "29.76")
  ("none?" "14.48")
  ("all?" "12.97")
  ("AND" "11.77")
  ("any?" "10.79")
  ("OR" "9.69")
  ("any" "9.14")
  ("loop" "8.96")
  ("all" "7.56")
  ("partition" "7.5")
  ("none" "6.33")
  ("pass" "4.72")
  ("sieve" "4.51")
  ("inverter" "4.21")
  ("amp" "4.1")
  ("NAND" "4.09")
  ("or%" "1.86")
  ("select" "1.6")
  ("or" "1.59")
  ("and%" "1.57")
  ("not" "1.5")
  ("NOR" 1)
  ("NOT" 1)
  ("XNOR" 1)
  ("XOR" 1)
  ("and" 1)
  ("apply" 1)
  ("block" 1)
  ("bundle" 1)
  ("catchall-template" 1)
  ("clos" 1)
  ("collect" 1)
  ("crossover" 1)
  ("currying" 1)
  ("effect" 1)
  ("fanout" 1)
  ("feedback" 1)
  (">>" 1)
  ("<<" 1)
  ("gate" 1)
  ("gen" 1)
  ("group" 1)
  ("if" 1)
  ("input aliases" 1)
  ("loop2" 1)
  ("one-of?" 1)
  ("relay" 1)
  ("relay*" 1)
  ("sep" 1)
  ("switch" 1)
  ("tee" 1)
  ("template" 1)
  ("thread" 1)
  ("try" 1)
  ("unless" 1)
  ("when" 1)
  ("esc" "0.68")
  ("ground" "0.58")
  ("thread-right" "0.56"))
```
Summary:
**23 slower
3 faster
35 same**

The criteria for displaying a relative performance factor other than 1 was r < 0.75x or r > 1.5x.

(This was done on the  on the latest `add-bindings-to-qi` branch)

benknoble commented:
> Summary:
> 
> **23 slower
> 
> 3 faster
> 
> 35 same**


To be fair, the one's that got slower mostly got slow by large factors (>=4, >=40)!

countvajhula commented:
Yeah it's just to give us an idea of progress rather than a report card. I'd love to get them all to <= 1 prior to merging this!

countvajhula commented:
Alrighty, we're getting the ball rolling on this again. I'll go ahead and merge #95 and then we can return to #76 and aim to Beat Racket™️  on one benchmark. That would bring us to the close of the compiler project where we can do the final review and merge this work, and that would then set us up for further optimizations over time on the main branch 🙂

countvajhula commented:
Hello everyone! You may have noticed that the `first-optimizations` branch was merged a few days ago (the actual one, this time!). So now, after more than a year of work, we're finally close to releasing this! 🏁 🏎️ 💨 And what's more, we're beating Racket on some benchmarks, as we set out to do!

We are targeting the solstice, i.e. **Dec 22**, as the release date for _Qi 4_ -- a little under two weeks -- and it coincides with a Qi meeting as it's on a Friday. If you have time to review this PR, that would be much appreciated. Since it's such a huge PR, here's a quick overview of things as a refresher -- you could probably also refer to [the meeting notes](https://github.com/drym-org/qi/wiki/Meeting-Notes) over the last year for additional context and the general story of how we got here.

Overview of new operation: At a high level, when we write a flow, it first expands that expression to the Qi core language (e.g. `switch` is rewritten to `if` as part of this process, analogously to how `cond` is rewritten to `if` as part of Racket expansion). Then, the core language expression is compiled to produce Racket code which is then executed by the underlying Racket interpreter. The compilation phase is itself done in several stages. First, the input expression is normalized to a standard form (["Qi Normal form"](https://github.com/drym-org/qi/wiki/Qi-Compiler-Sync-Sept-15-2023#qi-normal-form-and-proving-correctness)). Then, that normalized expression is optimized (currently the only optimization we perform is deforestation, for list-based functional pipelines). And finally, the optimized expression is translated into Racket ("code generation step", fulfilled by the `qi0->racket` macro).

Where is all this happening in the code, you might ask? Well, I've updated the [source code layout wiki](https://github.com/drym-org/qi/wiki/Source-Code-Overview) to include a comprehensive overview of every module in the source tree, for your reference.

You could try things locally if you like (after checking out this branch):

```
make build
make test-flow 
make test-compiler
make profile-competitive
```

If you have any Qi codebases (esp. @benknoble and @NoahStoryM ), now is a great time to test those against this compiler branch. There are some [known backwards incompatibilities](https://github.com/drym-org/qi/wiki/Qi-Compiler-Sync-Nov-4-2023), and if you happen to run into any unknown ones that would be good to know 😆 

Here are some specific items that may need review:

- Are there any inequivalencies in the normalization pass? Is there anything else it should be doing? Is it doing anything more than it should be?
- Any concerns with termination of compiler passes? Currently we run each pass until it reaches a fixed point. We've been implicitly assuming it will always get there, but we don't know that for sure.
- Any thoughts on the benchmarks we're using (in `qi-sdk/profile/nonlocal`)? Any ideas for other benchmarks?
- Do the existing optimizations look OK?
- Error messages, code layout, tests, coverage, anything else you can think of
- What do you think of the new syntax for bindings?

There are still a few items we'd like to do before release:

- documentation! Both user and developer docs.
- code coverage is currently showing as 72% but that doesn't seem accurate. We'd like to fix that and hopefully bring it up to 100% soon.
- bindings are not properly scoped with respect to the `switch` conditional form. We are considering this a blocker for release and hope to look into it soon (@michaelballantyne , we might need your help on it!)
- incorporating competitive benchmarks into the [live dashboard](https://drym-org.github.io/qi/benchmarks/)

That's all I can think of for the moment. I'll aim to keep the PR description up to date with outstanding items. Thanks! Let's get this compiler out into the wild! 🐺 🐅 🌲

countvajhula commented:
@benknoble If you can also look at normalization that'd be great! Last time you found a bunch of issues with it, soooo, I think it would give us more confidence if you signed off on it 😄 

I'm also curious how the switch-over goes with the install. IME it's been a matter of uninstalling any existing version of Qi (likely with `raco pkg remove --force qi`), and then `make install` within the local cloned Qi repo. After that, you'd need to `make build` each time you switch branches.

And re: merge conflicts, agreed, I'm planning to rebase this branch soon (I will give a heads up).

benknoble commented on qi-lib/flow/core/normalize.rkt:
```diff
@@ -0,0 +1,80 @@
+    [(thread _0 ... (pass f) (amp g) _1 ...)
+     #'(thread _0 ... (amp (if f g ground)) _1 ...)]
+    ;; merge amps in sequence
+    [(thread _0 ... (amp f) (amp g) _1 ...)
+     #'(thread _0 ... (amp (thread f g)) _1 ...)]
```
This is not generally sound (I'm not sure if this is the smallest example):
```
> (~> (3) (>< (-< add1 sub1)) (>< (-< sub1 add1)))
3
5
1
3
> (~> (3) (>< (~> (-< add1 sub1) (-< sub1 add1))))
sub1: arity mismatch;
 the expected number of arguments does not match the given number
  expected: 1
  given: 2
 [,bt for context]
```

benknoble commented on qi-lib/flow/core/normalize.rkt:
```diff
@@ -0,0 +1,80 @@
+    [(thread (~datum _) ...)
+     #'_]
+    ;; identity flows composed using a relay
+    [(relay (~datum _) ...)
+     #'_]
```
This is more permissive than current Qi:
```
> (~> (3) (relay _ _ _))
relay: arity mismatch;
 the expected number of arguments does not match the given number
  expected: 0
  given: 2
 [,bt for context]
```

benknoble commented on qi-lib/flow/core/normalize.rkt:
```diff
@@ -0,0 +1,80 @@
+    ;; collapse `values` and `_` inside a threading form
+    [(thread _0 ... (esc (#%host-expression (~literal values))) _1 ...)
+     #'(thread _0 ... _1 ...)]
+    [(thread _0 ... (~datum _) _1 ...)
+     #'(thread _0 ... _1 ...)]
```
I believe this is covered by a previous case on line 46.

benknoble commented on qi-lib/flow/core/deforest.rkt:
```diff
@@ -0,0 +1,415 @@
+  (define (prettify-flow-syntax stx)
+    (syntax-parse stx
+      #:datum-literals (#%host-expression esc #%blanket-template #%fine-template)
+      (((~literal thread)
```
I'd like to know why this `thread` has to be literal; in the rest of the compiler it's matched by datum.

benknoble commented:
> I'm also curious how the switch-over goes with the install. IME it's been a matter of uninstalling any existing version of Qi (likely with `raco pkg remove --force qi`), and then `make install` within the local cloned Qi repo. After that, you'd need to `make build` each time you switch branches.

I remembered the trick by checking the docs. I went to Qi clone and checked out the main branch (really the commit hash needs to match, or it needs to be possible to `pull --ff-only` to the right commit). Then `raco pkg update --clone . qi` (and respond `a` to the prompt) switched my install to my clone and re-`setup` dependent packages.

benknoble commented:
> 1. Do my tests pass (Frosthaven Manager and old Advent of Code)?

I fixed FH: https://github.com/benknoble/frosthaven-manager/commit/73a918d486a2fd5046e791cb5c83d127f8c93e9f (just another missed `esc`). I'm debugging some AoC tests… not sure what's happening there right now.

benknoble commented:
> 1. Do my tests pass (Frosthaven Manager and old Advent of Code)?

I needed the following patch to the AoC code to disable the unsound `amp`-to-`amp` optimization:
```patch
diff --git i/day10/solution.rkt w/day10/solution.rkt
index 4d2dec2..d4653c7 100644
--- i/day10/solution.rkt
+++ w/day10/solution.rkt
@@ -56,6 +56,7 @@ (define-flow part2*
   (~>> sep
        ;; needs to be separate stage from below because grounds some inputs
        (amp incomplete?)
+       (esc (λ xs (apply values xs)))
        (amp (~> sep (amp (hash-ref openers _)) score-repair))
        collect
        (median <)))
diff --git i/day17/solution.rkt w/day17/solution.rkt
index 99f47c5..33802c8 100644
--- i/day17/solution.rkt
+++ w/day17/solution.rkt
@@ -171,6 +171,7 @@ (define-flow string->xm-xM-ym-yM
   (~> (string-split "=") sep
       (select 2 3)
       (amp (~> (string-split "..") sep))
+      (esc (λ xs (apply values xs)))
       (amp (~> open-input-string read))))
 (define-flow file->xm-xM-ym-yM
   (~> file->string string->xm-xM-ym-yM))
diff --git i/day19/solution.rkt w/day19/solution.rkt
index 4c17335..b8f3504 100644
--- i/day19/solution.rkt
+++ w/day19/solution.rkt
@@ -24,6 +24,7 @@ (define/match (manhattan p1 p2)
 
 (define-flow (reld* ps)
   (~> (combinations 2) sep (amp (-< _ reverse))
+      (esc (λ xs (apply values xs)))
       (amp (~> sep (-< reld 1>) collect))
       collect))
```

Obviously this was terrible for performance. The following patch also worked and suggests a sound rewrite (though I'm not sure if it's desirable!):
```patch
diff --git i/day10/solution.rkt w/day10/solution.rkt
index 4d2dec2..93aadaf 100644
--- i/day10/solution.rkt
+++ w/day10/solution.rkt
@@ -55,8 +55,8 @@ (define-flow score-repair
 (define-flow part2*
   (~>> sep
        ;; needs to be separate stage from below because grounds some inputs
-       (amp incomplete?)
-       (amp (~> sep (amp (hash-ref openers _)) score-repair))
+       (amp (~> incomplete?
+                (amp (~> sep (amp (hash-ref openers _)) score-repair))))
        collect
        (median <)))
 (define-flow part2 (~> file->lines part2*))
diff --git i/day17/solution.rkt w/day17/solution.rkt
index 99f47c5..cb0e9de 100644
--- i/day17/solution.rkt
+++ w/day17/solution.rkt
@@ -170,8 +170,8 @@ (define (solve xm xM ym yM)
 (define-flow string->xm-xM-ym-yM
   (~> (string-split "=") sep
       (select 2 3)
-      (amp (~> (string-split "..") sep))
-      (amp (~> open-input-string read))))
+      (amp (~> (string-split "..") sep
+               (amp (~> open-input-string read))))))
 (define-flow file->xm-xM-ym-yM
   (~> file->string string->xm-xM-ym-yM))
 
diff --git i/day19/solution.rkt w/day19/solution.rkt
index 4c17335..fdc25d0 100644
--- i/day19/solution.rkt
+++ w/day19/solution.rkt
@@ -23,8 +23,8 @@ (define/match (manhattan p1 p2)
       (abs (- z1 z2)))])
 
 (define-flow (reld* ps)
-  (~> (combinations 2) sep (amp (-< _ reverse))
-      (amp (~> sep (-< reld 1>) collect))
+  (~> (combinations 2) sep (amp (~> (-< _ reverse)
+                                    (amp (~> sep (-< reld 1>) collect))))
       collect))
 
 (struct station [p beacons relds] #:transparent)
```

The error messages were fairly inscrutable, too:
```
raco test: (submod "day10/check.rkt" test)
raco test: @(test-responsible '(benknoble))
--------------------
ERROR
name:       check-equal?
location:   .../advent2021/check.rkt:39:9

list?: arity mismatch;
 the expected number of arguments does not match the given number
  expected: 1
  given: 0
--------------------
raco test: (submod "day17/check.rkt" test)
raco test: @(test-responsible '(benknoble))
--------------------
ERROR
name:       check-equal?
location:   .../advent2021/check.rkt:38:5

composed: arity mismatch;
 the expected number of arguments does not match the given number
  expected: 4
  given: 2
--------------------
--------------------
ERROR
name:       check-equal?
location:   .../advent2021/check.rkt:39:9

composed: arity mismatch;
 the expected number of arguments does not match the given number
  expected: 4
  given: 2
--------------------
raco test: (submod "day19/check.rkt" test)
raco test: @(test-responsible '(benknoble))
--------------------
ERROR
name:       check-equal?
location:   check.rkt:5:2

list?: arity mismatch;
 the expected number of arguments does not match the given number
  expected: 1
  given: 2
--------------------
4/5 test failures
```
```
list?: arity mismatch;
 the expected number of arguments does not match the given number
  expected: 1
  given: 0
  context...:
   .../core/compiler.rkt:333:9
   /Users/Knoble/code/qi/qi-lib/flow/core/impl.rkt:119:0: ~map
   /Users/Knoble/code/qi/qi-lib/flow/core/impl.rkt:127:0: map-values
   .../racket/cmdline.rkt:191:51
   body of (submod "/Users/Knoble/code/advent2021/day10/solution.rkt" main)
cpu time: 12 real time: 12 gc time: 5
344193
composed: arity mismatch;
 the expected number of arguments does not match the given number
  expected: 4
  given: 2
  context...:
   .../racket/cmdline.rkt:191:51
   body of (submod "/Users/Knoble/code/advent2021/day17/solution.rkt" main)
list?: arity mismatch;
 the expected number of arguments does not match the given number
  expected: 1
  given: 2
  context...:
   .../core/compiler.rkt:333:9
   /Users/Knoble/code/qi/qi-lib/flow/core/impl.rkt:119:0: ~map
   /Users/Knoble/code/qi/qi-lib/flow/core/impl.rkt:127:0: map-values
   .../core/compiler.rkt:394:9
   [repeats 1 more time]
   body of (submod "/Users/Knoble/code/advent2021/day19/solution.rkt" main)
```

countvajhula commented:
@benknoble Thank you, I knew you were the right person for the job 😆 . I've addressed those normalization issues and added them to the "counterexamples" test suite in #133.

Re: the error messages, those are indeed pretty bad. The `list? arity` error is at least documented as a [commonly encountered error](https://docs.racket-lang.org/qi/Field_Guide.html#%28part._.List_.Arity_.Mismatch%29) (but it would be better to not expose the implementation here, of course), while all of the "`composed`" errors are likely to be fixed by @NoahStoryM 's changes in #73 , but I think we ran into some blockers there -- any updates on that @NoahStoryM ?

benknoble commented:
One day I'd like to try formalizing the semantics a bit and provide a way to test or prove that optimizations are valid... maybe PLT Redex will be a good fit for this, but I have very little experience with it (I didn't get too far into the tutorial because something broke for me and I never got an answer on how to proceed.)

countvajhula commented:
@benknoble That would be amazing and btw @dzoep has also expressed interest in that, and was considering submitting a [research proposal](https://github.com/drym-org/qi/wiki/Qi-Compiler-Sync-Sept-15-2023#qi-normal-form-and-proving-correctness) for it a little while back 🤓 . Can't wait to see what these lead to.

NoahStoryM commented:
> while all of the "composed" errors are likely to be fixed by @NoahStoryM 's changes in https://github.com/drym-org/qi/pull/73 , but I think we ran into some blockers there -- any updates on that @NoahStoryM ?

Hi @countvajhula,

These PRs are quite old, and I recalled their changes. #70 and #73 are related, and I summarized their changes as follows:

#70 mainly did these works:

1. Some operators always create the same function, such as `(☯ (gen))`, `(☯ ⏚)`, `(☯ (-<))`:
> (eq? (☯ ⏚) (☯ ⏚))
#f
> (eq? (☯ (gen)) (☯ (gen)))
#f
> (eq? (☯ (-<)) (☯ (-<)))
#f

#70 made them return the same function.

2. `==*` and `-<` can be regarded as monoids, so the identity elements in the parameters can be filtered out before performing the calculation.

3. There are many places in qi that directly return `(lambda args body)`, this PR named them: `(let ([compiled-fname-flow (lambda args body)]) compiled-fname-flow)`.

Because these works were overlapping parts between these changes, I didn’t split them into multiple PRs at the time.


#73’s work is like this: when running `(define f (☯ ...))`, `f`'s object-name is usually `composed` or `compiled-fname-flow`, #73 uses `procedure-rename` to rename its name to `f`.

But I found that `procedure-rename` would cause some problems: https://github.com/racket/racket/issues/4455, so it is not suitable to merge #73 before this issue is fixed.


I think qi has changed a lot in the past year, and I'm willing to deal with the PRs I left before `lets-write-a-qi-compiler` is merged. (I guess I should merge these PRs into the `lets-write-a-qi-compiler` instead of `main`?)

countvajhula commented:
@NoahStoryM I appreciate the summary! I'm not sure I'd consider most of those changes to be within the scope of the compiler work, but if you create PRs based on the compiler branch (you may want to wait a few days since I am planning to rebase this branch to `main`), we can likely aim to discuss merging into main after the compiler branch is merged. I admit that as I recall, I either wasn't convinced or didn't have enough context to accept all of the changes, so if you could make the PRs narrowly scoped, that might help to focus our discussions and understand the proposals better. Thank you for your patience! 🙏

NoahStoryM commented:
I understand your concerns, and I’m willing to wait until the rebase is done, and then split my previous changes into smaller PRs. 😊

countvajhula commented:
Hi all, I'm going to rebase this branch now. If you have any WIP against it and would prefer to avoid merge conflicts, let me know and I can wait. Otherwise, you'll just need to rebase your WIP branch and it should be fine.

countvajhula commented:
OK the branch is rebased to main and all the tests are passing now. Initially there was a failing test, but after further investigation, I concluded that the test was invalid to begin with, as I think we agreed some time ago that we would not support keyword arguments as flow inputs! (Yet, I am the one who wrote this invalid test ...). Technically this does represent a backwards incompatibility as it apparently worked on main before (even though I believe unintentionally).

Interestingly enough @dzoep , the coverage is now reporting on CI as 95%.... maybe the recent change from `module+ main` to `module+ test` was needed for it to pick up the tests!

countvajhula commented:
Alright folks, we're almost there! There are just a few items remaining in outstanding work, and I've updated the list in the PR description to include everything I can think of. I've also included in box brackets at the end of each item who I think might be interested in working on that item. If you'd like to do it, just update the description to change `[Name/Name ...?]` to `[Name]` on that item 😄 Otherwise if it is unclaimed, whoever gets to it first can just do it. Thank you!

countvajhula commented:
Hey all, heads up: I plan to rebase the compiler branch today to get the latest documentation changes from the main branch.

countvajhula commented:
Hi all, we're getting close! Now is a great time for another round of review in advance of our planned release on Jan 12. Any thoughts, comments, feedback welcome. Thanks 😊

countvajhula commented:
Rebasing this branch now...

michaelballantyne commented on qi-lib/flow/core/private/form-property.rkt:
```diff
@@ -0,0 +1,49 @@
+;; Whenever we transform syntax we need to propagate this property too,
+;; so that subsequent optimization passes see it. We also need to attach
+;; this property in tests.
+(define (tag-form-syntax stx)
```
This is a dangerous operation only appropriate to use in tests because it indiscriminately tags every node in a syntax object as a form. This of course would lead to the incorrect optimizations we originally added the property to correct if this operation were applied to syntax in the real compiler. The comment for this form should clarify the intended use and danger of misuse.

countvajhula replied:
Paraphrased and used your comment there.

michaelballantyne commented on qi-lib/flow/core/pass.rkt:
```diff
@@ -0,0 +1,71 @@
+  ;; collide with a user-defined #%host-expression binding, but that
+  ;; would never be hit in practice since that would be rewritten
+  ;; through expansion to a use of the core language. In general,
+  ;; we should be using ~literal matching throughout the compiler.
```
Now that we rely on the syntax property to only apply transformations at proper form positions, I don't think we need to use `~literal` matching anymore. I think the use of `~literal` matching was an early attempt to avoid misinterpreting binding positions before we realized that other overlaps could happen like the `(partition [tee ...])` case and thus that we needed a different strategy.

michaelballantyne commented on qi-lib/flow/core/deforest.rkt:
```diff
@@ -0,0 +1,405 @@
+#lang racket/base
+
```
Deforestation is by far the most difficult to understand part of this PR for me. I think this file deserves a big 100-line introductory comment explaining the compilation strategy, maybe with an example, and the structure of the code that follows. If there is already an up-to-date description on the wiki a link there would be okay. But as is I'm not sure this file is maintainable other than by Dominik without a lot of effort. I have not tried to review it in detail.

countvajhula replied:
Dominik did document this in the wiki here: https://github.com/drym-org/qi/wiki/The-Compiler#optimizations . I haven't reviewed it in detail so I confess I'm not sure re: maintainability, but on first glance it seems comprehensive. Some notes though:

- That wiki doesn't include examples and I agree they would be nice to have
- I second that it would be useful to write a comment at the top of the module explaining the structure of the code that follows, and maybe even signpost each section, like ";; Core stream implementations" ";; Syntax classes to match optimizable input patterns"
- I don't immediately recall exactly what `make-fine-curry` and `make-producer-curry` do -- the comments explain that they make curry 🍛  but they don't say for what purpose or where it's used. 

This module is pretty dense and could benefit from more signposting/putting the code in context. Wdyt @dzoep ?

For now I'll go ahead and add a link to the wiki at the top of this source module (will tack it on to the PR that fixes the marshalling issue).

michaelballantyne replied:
The wiki page covers what I was looking for, and with a link from the code I'm happy!

samdphillips replied:
The source may out-live a Github Wiki so moving the explanation to the implementation may be desirable.

countvajhula replied:
Yeah I'm not really sure what the best thing here is. It would be nice for the source files to be as self-contained as possible, and I definitely see the possibility of source and documentation diverging if they are kept separate, presenting a nonlocal relationship to be maintained.

On the other hand, there's value in such didactic documentation being accessible in "book" form without having to read source code, and there may be a balance to be struck.

Options:

- have the docs only on the wiki or only in the source
- duplicate docs in both places
- have implementation-specific docs (e.g. `generate-fused-operation`) in the source, and technical and theoretical material (e.g. what is stream fusion, here are some resources, here's how we use it at a high level in terms of e.g. producers transfomers and consumers, here where this is reflected in the source code)

The third option will likely entail some overlap between the docs, but as they reflect a different orientation it's possible the overlap would not be very significant (and they could mutually delegate to one another). If we think this is a good option, we could take the existing docs from the compiler wiki and distribute it along these lines.

Re: outlive, we could culturally commit to maintaining a wiki alongside the code, and augment our processes to support that in a sustainable way. E.g. we could add a bullet point in the PR template to say something like, "Have you updated the wiki (if applicable) to reflect the changes in your PR?"

We would still need to have links to the wiki if we went with this option, and I'm not sure how we'd avoid actual URLs, and actual references to GitHub in that case...

Related: #139

benknoble replied:
The Git project maintains a glossary and other technical contribution material as part of the repository. They use plain-text, but scribble would do nicely and could be hosted as developer docs (perhaps as part of the qi-sdk package?).

countvajhula replied:
That sounds like a good option. From a wiki perspective, it would be ideal if (1) we had more styles for Scribble that we could customize to give it a wiki-like feel (e.g. "edit" links like those on the [Python Developer's Guide](https://devguide.python.org/developer-workflow/lang-changes/#langchanges)), (2) it recognized simplified wiki-style links resembling `[[a]]` and `[[a|b]]`, and (3) if we could edit and preview it via a web UI (e.g. Wikipedia/Github wiki, maybe Readthedocs). If we could have at least some of these that would make Scribble even more compelling vs Markdown/RST for this purpose. But I really don't know what the level of effort on these would be. Would it make sense to use Pollen?

benknoble replied:
Section titles should have clickable edit links. Otherwise maybe you just want a collection of Markdown documents?

countvajhula replied:
Interesting, you mean this, when we click the heading?

```
Link to this section with
 @secref["Using_Qi_from_the_Host_Language"
         #:doc '(lib "qi/scribblings/qi.scrbl")]
Document source
 https://github.com/drym-org/qi/tree/main/qi-doc/scribblings/qi.scrbl
```

I forgot about this and it's quite functional. I think it's almost there, but it would be better if it was further simplified / specialized to a self-contained wiki, where there would be an explicit [Edit] button that simply follows the link to the Scribble source file, and where "link to this section" is unnecessary since we need not support cross-linking to other docs (so all links, except HTTP URLs, are internal, allowing the link syntax to be simple like [[a]] or [[a|b]]). I feel we especially want the developer docs to be [convenient and fast to edit](https://github.com/drym-org/qi/wiki/Writing-Developer-Docs).

We might be looking for a Scribble-based language here, comprising a set of macros/reader extensions + a CSS file. "Scribwiki"? 😆 But I'd love to learn more about Pollen to see if it's already usable in this way.

I'm also not sure if we'd want to move the entire wiki to Scribble or only the Developer's Guide.

countvajhula replied:
Also the link to the source file is just the root scribble file and not the specific one containing the content. Is that a bug?

benknoble replied:
> Also the link to the source file is just the root scribble file and not the specific one containing the content. Is that a bug?

I think it's a regrettable consequence of `include-section`. Let me ask around.

countvajhula commented on qi-doc/scribblings/field-guide.scrbl:
```diff
@@ -8,18 +8,24 @@
                     racket]]
 
 @(define eval-for-docs
```
There wasn't a specific reason, but when I tried to change it to use a single evaluator, it was initially getting some unexpected results because I think the evaluator was stateful across Scribble modules. I've now made it so that all of the modules use a common "factory" for creating the evaluator they use.

countvajhula commented on qi-lib/flow/core/debug.rkt:
```diff
@@ -0,0 +1,22 @@
+;; giving us visibility into this process for debugging purposes.
+;; Note that this currently does not distinguish substeps
+;; of a parent expansion step.
+(define-syntax-rule (qi-expansion-step name stx0 stx1)
```
Changed it to a function and it's no longer exported.

countvajhula commented on qi-lib/macro.rkt:
```diff
@@ -88,32 +87,26 @@
+     #:with (spaced-form-name ...) (map introduce-qi-syntax
                                         (attribute form-name))
      #'(begin
          (define-syntax spaced-form-name (make-qi-foreign-syntax-transformer #'form-name))
```
Made these changes.

countvajhula commented:
Hi folks, if you're still planning to review this, please get your reviews in by tomorrow. If there are no further blockers, we will aim to merge this on Friday morning PST 😁 . At this stage, if you have comments that aren't blockers, they may be addressed after release, but they are still appreciated. Thanks!

countvajhula commented:
We don't anticipate further code changes before tomorrow's release, so now is a great time to test against this branch with [Frosthaven](https://github.com/benknoble/frosthaven-manager/) (@benknoble ), [Qi-Cat](https://github.com/NoahStoryM/qi-cat) (@NoahStoryM ), [Relation](https://github.com/countvajhula/relation) (@countvajhula ) and [Qi-Circuit](https://github.com/chansey97/qi-circuit) (@chansey97 ). For instructions on how to test against a specific repo branch, [this comment](https://github.com/drym-org/qi/pull/74#issuecomment-1852355343) (and the quoted text) may be helpful. Thanks! 😄

benknoble commented:
I had to pass `--clean` to raco setup when building this branch recently using `make build-all`, which spat out the following deps warnings:
```
raco setup: unused dependencies detected
  for package: "qi-lib"
  on packages:
   "syntax-spec-v1"
   "fancy-app"
   "macro-debugger"
raco setup: unused dependencies detected
  for package: "qi-test"
  on packages:
   "adjutor"
   "math-lib"
   "syntax-spec-v1"
   "qi-lib"
   "rackunit-lib"
   "testing-util-lib"
raco setup: unused dependencies detected
  for package: "qi-sdk"
  on packages:
   "cover-coveralls"
   "adjutor"
   "relation-lib"
   "math-lib"
   "qi-lib"
   "csv-writing"
   "cover"
   "cli"
   "collections-lib"
   "require-latency"
raco setup: unused dependency detected
  for package: "qi-probe"
  on package:
   "qi-lib"
raco setup: unused dependencies detected
  for package: "qi-doc"
  on packages:
   "scribble-abbrevs"
   "qi-probe"
   "qi-lib"
   "scribble-html-lib"
   "racket-doc"
   "sandbox-lib"
   "scribble-lib"
   "scribble-math"
   "metapict"
```

I can confirm that, after also `raco setup frosthaven-manager advent2021`, the tests `raco test -xc -j 8 frosthaven-manager advent2021` passed. If I find any bugs while running the FHM GUI, I'll let you know ASAP (we play weekly, so earliest you'll hear is next week).

benknoble commented:
Ah, please ignore the dependency warnings. I misunderstood the output of `--clean`.
